apps/designer/app/routes/$.tsx

```diff
@@ -8,7 +8,7 @@ import { useLoaderData } from "@remix-run/react";
 import { InstanceRoot, Root } from "@webstudio-is/react-sdk";
 import { loadCanvasData, type CanvasData } from "~/shared/db";
 import config from "~/config";
-import { db } from "@webstudio-is/project/index.server";
+import { db } from "@webstudio-is/project/server";
```

Comment:
This actually still references `something.server` because this `server` file just reexports `index.server.js`. However, I've retested this - and it seems that this part doesn't work now.

I think the fix here is more important than this Remix safeguard. I could investigate separately how to make this work later on though. I think though that maybe it would be better to introduce a throwing expression there that would work regardless of the Remix convention, what do you think? It would still prevent the modules from being imported by the client code.

Reply:
If after we merge this PR what was previously server only code will appear in client side bundles we probably should solve this before merging, or maybe I misunderstood something.

> I think though that maybe it would be better to introduce a throwing expression there that would work regardless of the Remix convention, what do you think? It would still prevent the modules from being imported by the client code.

Not sure I understand. Do you mean throwing an error in browser context so the code wouldn't run in the browser?

Reply:
Maybe we should actually just do `import { db } from "@webstudio-is/project";` everywhere, and remove these entry points. 

It should just work because for the client bundles remix does not import loaders and actions functions so they will be tree-shaken and all imports that they use along with them.

Remix docs say that `.server` shouldn't be necessary usually.

And for the rare cases when it's necessary, I think a better approach is to create a `*.server` file in the app rather than in a package.

Reply:
@kof do you know why we started to import from `.server` rather than normally? Is it to be extra safe or have we actually had a problem of backend code appearing in the browser?

Reply:
its an actual problem once you end up importing in client side server stuff, remix will crash at you

Reply:
creating .server files in the app instead of a package is indeed an option, as long as contributors to the package always get the error if that change can't be used in the app

Reply:
I've pulled this PR locally, run `yarn dev`, and got this error in the browser: <img width="1147" alt="Screenshot 2022-11-05 at 18 09 37" src="https://user-images.githubusercontent.com/825702/200123947-49df0ea4-7fc0-4495-8769-d6564e526fd3.png">

So it seems `import ... from ".../server"` is not the same as `import ... from ".../something.server"` for Remix.

I guess we have two options:

 1. Remove `/server` entries and put all exports in `index.ts` normally. In the app import via intermediary `*.server.ts` files
 2. Just rename these `server.js` files to `index.server.js` so the imports would look like `import ... from ".../index.server"`
 
I like `#1` a bit better but no strong opinion, and not sure `#2` will work.

Reply:
> If after we merge this PR what was previously server only code will appear in client side bundles we probably should solve this before merging, or maybe I misunderstood something.

Nothing should change - but the risk of introducing a server-related import to the client bundle increases here. I will take a second look into how Remix actually implements their `.server` convention and maybe I will be able to do smth about it.

> Not sure I understand. Do you mean throwing an error in browser context so the code wouldn't run in the browser?

Yes, smth like that. With that, it wouldn't rely on the Remix convention and it would work in the same way "everywhere".

> It should just work because for the client bundles Remix does not import loader and action functions so they will be tree-shaken and all imports that they use along with them.

Unfortunately, it's not always that simple and thus relying on this introduces some new risks.

> And for the rare cases when it's necessary, I think a better approach is to create a *.server file in the app rather than in a package.

That I agree with - `*.server` convention feels mostly relevant for the app code. Packages should be more reusable and should work outside of Remix etc. That being said - sometimes it's useful to have different entries per environment. For example, React comes with `react-dom` and `react-dom/server` (and even `react-dom/client` that has been added in React 18).

My first instinct would be to say that:
- you either have a lib that works in any env and you only provide a single entrypoint 
- or you have a lib that needs different implementation per environment but it still works in all~ of them and then u can have a single entrypoint but you need to alias that entrypoint per env (nowadays this gets easier with `package.json#exports`)
- or you have a library that only works in a subset of the available envs. You can have only a single entrypoint there as importing it in an incompatible env would crash one way or another so you probably would notice it
- or you have a library where a subset of the API only works in some specific env. In this case, I recommend having multiple entrypoints (like in this PR).

What you probably don't want to have is a single entrypoint in which some APIs are isomorphic but some APIs are env specific. I would classify this as a code smell. I mean that this probably ain't a good idea:
```js
import { worksEverywhere, onlyWorksInNode } from 'pkg'
```

Of course, there are no hard rules here and you might sometimes break out of the listed scenarios and still be fine... but the line is blurry and it gets harder to control this. The mentioned scenarios are somewhat straightforward and I would stick to them. It's especially important not only because of the imported APIs but because of what they might depend on.

The mentioned tree-shaking scenario and isomorphic files like in the case of Remix blur the line BUT the "mixed" entrypoint can still be a problem. Tree-shaking (with `sideEffects: false` in the imported packages) might perform much better if a file has a mixed client/server code while not having mixed imports - if imports are either isomorphic, client-only, or server-only, you can more safely rely on bundling and tree-shaking.

> cannot initialize routeModules

Hmm, I've seen this before some of my fixes but I've thought that I already got rid of this. Thanks for checking and I gonna take another pass at this.

Note that in this case, it might be unrelated to the `.server` convention. I got this exact error before simply because one client-only chunk had a `ReferenceError` in it and it failed to initialize some global variables used by Remix for rehydration. But if this still happens I need to recheck why this happens now.

Reply:
The problem is that when writing code in the package, you don't know if its being used serverside. If we can show the user rightaway that the change is now going to bring the server-side code to the client - then we may be able to get away with it.

It has to crash the app when user is bringing server code to the client. If we can do that, we can keep .server to the app

Reply:
Another big worry of mine is that client code can get stuff in that its not supposed to get just because we write the code in packages in a way that can't be treeshaken and in addition there is no .server hint.

I am super worried about growing bundle size for the client and it has as of right now already big problems we need to investigate. It is way bigger than it should be.

Reply:
It may have to do with the fact that we have canvas and parent document and we load twice the things that are actually the same and could be taken from browser cache. I am not sure.

I bet our bundler can't see the difference between iframe stuff and main window: e.g. see https://webstudio-designer-7humw6tf0-webstudio-is.vercel.app/designer/5e1de84c-25c6-4cfb-913f-280c295a4199

half of the files are the same, visible by the size right away. So we probably 1/2 from 1.5mb is unnecessary

Reply:
Just mentioning this to make you aware, maybe there is a way to handle all of this better, but maybe its completely unrelated

Reply:
> It has to crash the app when user is bringing server code to the client. If we can do that, we can keep .server to the app

Ye, so with the `throw`-based idea this should very much be possible. 

> Another big worry of mine is that client code can get stuff in that its not supposed to get just because we write the code in packages in a way that can't be treeshaken and in addition there is no .server hint.

Isn't this always the risk if you rely on tree-shaking? I mean, you don't truly control what gets included and what is not. Using a single package can bring megabytes to your bundles. There is no granular control over that and the best mitigation is to monitor your bundles, your dependencies, and sometimes improve your own module graph if needed.

I might be confusing what do you have in mind though. Could you define "client code can get stuff in that its not supposed to get"? Are you purely concerned here from the bundlesize perspective? Or do you mean potential runtime crashes etc?

> It may have to do with the fact that we have canvas and parent document and we load twice the things that are actually the same and could be taken from browser cache. I am not sure.

This could be mitigated using multiple entrypoints. The bundler should be able to share common chunks between multiple entries. I'm not sure if this is supported by Remix and esbuild in the way that I know and understand this. It would very much be possible (and the best solution imho) in the webpack/rollup world. I can only assume that the same should be supported here too. To assess this I would have to take a closer look at your setup related to this stuff.

Reply:
I think that the fact that your `pkg/index.server` worked was a little bit accidental. The comment in the Remix's source code suggests that they only want to allow this convention for the files in the `/app` directory and they even mention there that it should be allowed for node modules to use `.server.js` with different semantics:
https://github.com/remix-run/remix/blob/cb7416beb87f2f008f05bafb0c25d263bf59851f/packages/remix-dev/compiler/plugins/emptyModulesPlugin.ts#L18-L26

The problem~ with their implementation is that they simply call `path.resolve(args.resolveDir, args.path)` and that doesn't resolve module specifiers. So the "resolved" path looks something like:
```
~/Code/webstudio-designer/apps/designer/app/routes/designer/@webstudio-is/project/index.server
```
and that satisfies their check. Should it though? I don't quite think so - I think this could even be reported as a bug to the Remix team.

I **think** that I've managed to fix most of the issues by reverting some of the changes introduced here previously and by introducing `sideEffects: false` to your packages. Thanks to this flag unused imports in your isomorphic files might be dropped entirely by the bundler (`import { worksEverywhere, onlyWorksInNode } from 'pkg'` might still be problematic but conceptually you try to stay away from exporting things like that so you should be fine).

I still think that you:
- either shouldn't bundle your monorepo packages on the server
- or bundle everything for the server.

This might be a separate discussion though so I've removed those changes from this PR and I hope that this will now work as expected.

Note that you are still using esm-only `nanoid` and this might create problems (just like any esm-only package consumed from your CJS bundle). But let's discuss and solve this separately later.